x-pack/plugins/security/public/management/roles/edit_role/edit_role_page.tsx

```diff
@@ -0,0 +1,569 @@
+import {
+  EuiButton,
+  EuiButtonEmpty,
+  EuiFieldText,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiForm,
+  EuiFormRow,
+  EuiPanel,
+  EuiSpacer,
+  EuiText,
+  EuiTitle,
+} from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import { FormattedMessage } from '@kbn/i18n/react';
+import React, {
+  ChangeEvent,
+  Fragment,
+  FunctionComponent,
+  HTMLProps,
+  useEffect,
+  useRef,
+  useState,
+} from 'react';
+import {
+  Capabilities,
+  FatalErrorsSetup,
+  HttpStart,
+  IHttpFetchError,
+  NotificationsStart,
+} from 'src/core/public';
+import { IndexPatternsContract } from '../../../../../../../src/plugins/data/public';
+import { Space } from '../../../../../spaces/common/model/space';
+import { Feature } from '../../../../../features/public';
+import {
+  KibanaPrivileges,
+  RawKibanaPrivileges,
+  Role,
+  BuiltinESPrivileges,
+  isReadOnlyRole as checkIfRoleReadOnly,
+  isReservedRole as checkIfRoleReserved,
+  copyRole,
+  prepareRoleClone,
+  RoleIndexPrivilege,
+} from '../../../../common/model';
+import { ROLES_PATH } from '../../management_urls';
+import { RoleValidationResult, RoleValidator } from './validate_role';
+import { DeleteRoleButton } from './delete_role_button';
+import { ElasticsearchPrivileges, KibanaPrivilegesRegion } from './privileges';
+import { ReservedRoleBadge } from './reserved_role_badge';
+import { SecurityLicense } from '../../../../common/licensing';
+import { UserAPIClient } from '../../users';
+import { DocumentationLinksService } from '../documentation_links';
+import { IndicesAPIClient } from '../indices_api_client';
+import { RolesAPIClient } from '../roles_api_client';
+import { PrivilegesAPIClient } from '../privileges_api_client';
+
+interface Props {
+  action: 'edit' | 'clone';
+  roleName?: string;
+  indexPatterns: IndexPatternsContract;
+  userAPIClient: PublicMethodsOf<UserAPIClient>;
+  indicesAPIClient: PublicMethodsOf<IndicesAPIClient>;
+  rolesAPIClient: PublicMethodsOf<RolesAPIClient>;
+  privilegesAPIClient: PublicMethodsOf<PrivilegesAPIClient>;
+  docLinks: DocumentationLinksService;
+  http: HttpStart;
+  license: SecurityLicense;
+  spacesEnabled: boolean;
+  uiCapabilities: Capabilities;
+  notifications: NotificationsStart;
+  fatalErrors: FatalErrorsSetup;
+}
+
+function useRunAsUsers(
+  userAPIClient: PublicMethodsOf<UserAPIClient>,
+  fatalErrors: FatalErrorsSetup
+) {
+  const [userNames, setUserNames] = useState<string[] | null>(null);
+  useEffect(() => {
+    userAPIClient.getUsers().then(
+      users => setUserNames(users.map(user => user.username)),
+      err => fatalErrors.add(err)
+    );
+  }, [fatalErrors, userAPIClient]);
+
+  return userNames;
+}
+
+function useIndexPatternsTitles(
+  indexPatterns: IndexPatternsContract,
+  fatalErrors: FatalErrorsSetup
+) {
+  const [indexPatternsTitles, setIndexPatternsTitles] = useState<string[] | null>(null);
+  useEffect(() => {
+    indexPatterns.getTitles().then(setIndexPatternsTitles, err => fatalErrors.add(err));
```

Comment:
Rather than using a fatal error here, we could instead show a toast notification. That would allow us to resolve https://github.com/elastic/kibana/issues/51756. When we fail, returning an empty array should allow the app to continue functioning.

Reply:
Do you think we should make a distinction between 403 and the rest?

Reply:
I'm torn -- checking for the `403` is more intentional, but on the other hand, this request isn't required for the page to function...

Let's check for the `403`, and use a fatal error for other failure codes. We'll have a better chance of catching a regression this way, instead of silently failing when we expected a successful response.

Reply:
:+1: